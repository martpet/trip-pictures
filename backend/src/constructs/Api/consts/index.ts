export * from './authorizationHeader';
export * from './lambdaEdgeViewerEvent';
