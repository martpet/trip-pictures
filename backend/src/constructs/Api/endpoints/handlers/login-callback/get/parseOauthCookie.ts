import {
  APIGatewayProxyEventV2,
  OauthCookieProps,
  parseEventCookies,
} from 'lambda-layer';

export const parseOauthCookie = (event: APIGatewayProxyEventV2) => {
  try {
    const { oauth } = parseEventCookies<'/login-callback'>(event);
    if (!oauth) throw new Error();
    return JSON.parse(oauth) as Partial<OauthCookieProps>;
  } catch (e) {
    console.error(e);
    return {};
  }
};
