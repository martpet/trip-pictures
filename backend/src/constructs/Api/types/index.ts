export * from './Cookies';
export * from './EnvVars';
export * from './LambdaEdgeOriginRequestHandler';
export * from './LambdaEdgeResponse';
export * from './OauthCookieProps';
export * from './OauthTokens';
export * from './RouteHeaders';
