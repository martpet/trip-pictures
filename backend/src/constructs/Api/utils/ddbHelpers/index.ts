export * from './ddbBatchGet';
export * from './ddbBatchWrite';
export * from './ddbScan';
