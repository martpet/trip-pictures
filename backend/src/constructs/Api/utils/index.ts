export * from './cookieName';
export * from './ddbHelpers';
export * from './edgeErrorResponse';
export * from './errorResponse';
export * from './getIdTokenPayload';
export * from './parseEventCookies';
export * from './parseLambdaEdgeEventCookies';
export * from './testHelpers';
