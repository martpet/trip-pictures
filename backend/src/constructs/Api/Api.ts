import { HttpApi } from '@aws-cdk/aws-apigatewayv2-alpha';
import { Fn, NestedStack, RemovalPolicy } from 'aws-cdk-lib';
import { Distribution } from 'aws-cdk-lib/aws-cloudfront';
import { HttpOrigin } from 'aws-cdk-lib/aws-cloudfront-origins';
import { ARecord, RecordTarget } from 'aws-cdk-lib/aws-route53';
import { CloudFrontTarget } from 'aws-cdk-lib/aws-route53-targets';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

import { Auth, Photos, Zone } from '~/constructs';
import { apiSubdomain, appEnvs } from '~/consts';
import { getCdkEnv } from '~/utils';

import { createDistroBehaviors } from './endpoints/createDistroBehaviors';
import { createRoutes } from './endpoints/createRoutes';

type Props = {
  zone: Zone;
  auth: Auth;
  photos: Photos;
};

export class Api extends NestedStack {
  constructor(scope: Construct, { zone, auth, photos }: Props) {
    super(scope, 'Api');

    const envName = getCdkEnv();
    const { appDomain } = appEnvs[envName];
    const apiDomain = `${apiSubdomain}.${appDomain}`;
    const api = new HttpApi(this, 'http-api');
    const origin = new HttpOrigin(Fn.select(1, Fn.split('://', api.apiEndpoint)));

    const logBucket = new Bucket(this, 'Distro-Log-Bucket', {
      removalPolicy: RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
    });

    const distribution = new Distribution(this, 'Distro', {
      defaultBehavior: { origin },
      domainNames: [apiDomain],
      certificate: zone.certificate,
      enableLogging: true,
      logIncludesCookies: true,
      logBucket,
    });

    createRoutes({ scope, api, auth, photos });

    createDistroBehaviors({
      scope,
      api,
      auth,
      distribution,
      origin,
    });

    new ARecord(this, 'A-Record', {
      zone: zone.hostedZone,
      recordName: apiSubdomain,
      target: RecordTarget.fromAlias(new CloudFrontTarget(distribution)),
    });
  }
}
