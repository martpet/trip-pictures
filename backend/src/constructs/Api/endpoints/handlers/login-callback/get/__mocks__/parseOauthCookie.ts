import { OauthCookieProps } from '~/constructs/Api/types';

export const dummyOauthCookieProps: OauthCookieProps = {
  stateNonce: 'dummyState',
  idTokenNonce: 'dummyIdTokenNonce',
  codeVerifier: 'dummyCodeVerifier',
};

export const parseOauthCookie = vi
  .fn()
  .mockName('parseOauthCookie')
  .mockReturnValue(dummyOauthCookieProps);
