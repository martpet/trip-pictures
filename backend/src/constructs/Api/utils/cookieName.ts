import { AllCookies } from '~/constructs/Api/types';

export const cookieName = (name: keyof AllCookies) => name;
