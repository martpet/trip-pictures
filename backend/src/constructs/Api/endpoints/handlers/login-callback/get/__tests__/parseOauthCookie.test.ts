import { itCalls, itReturns, parseEventCookies } from '~/constructs/Api/utils';

import { parseOauthCookie } from '../parseOauthCookie';

vi.mock('~/constructs/Api/utils/parseEventCookies');

const args = [
  {
    cookies: ['dummyCookie1', 'dummyCookie2'],
  },
] as unknown as Parameters<typeof parseOauthCookie>;

vi.mocked(parseEventCookies).mockReturnValue({
  oauth: JSON.stringify({ dummyOauthCookieContentKey: 'dummyOauthCookieContentValue' }),
});

describe('parseOauthCookie', () => {
  itReturns(parseOauthCookie, args);
  itCalls(parseEventCookies, parseOauthCookie, args);

  describe('when "oauth" cookie is missing', () => {
    beforeEach(() => {
      vi.mocked(parseEventCookies).mockReturnValueOnce({});
    });
    itReturns(parseOauthCookie, args);
  });

  describe('when "oauth" cookie value is not json-parsable', () => {
    beforeEach(() => {
      vi.mocked(parseEventCookies).mockReturnValueOnce({ oauth: 'non json value' });
    });
    itReturns(parseOauthCookie, args);
  });
});
