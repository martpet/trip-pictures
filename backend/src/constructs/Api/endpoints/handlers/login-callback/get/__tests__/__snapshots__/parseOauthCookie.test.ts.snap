// Vitest Snapshot v1

exports[`parseOauthCookie > calls "parseEventCookies" with correct args 1`] = `
[
  [
    {
      "cookies": [
        "dummyCookie1",
        "dummyCookie2",
      ],
    },
  ],
]
`;

exports[`parseOauthCookie > returns a correct value 1`] = `
{
  "dummyOauthCookieContentKey": "dummyOauthCookieContentValue",
}
`;

exports[`parseOauthCookie > when "oauth" cookie is missing > returns a correct value 1`] = `{}`;

exports[`parseOauthCookie > when "oauth" cookie value is not json-parsable > returns a correct value 1`] = `{}`;
